TykTechnologies/tyk
# TT-4354 Go plugins to support multiple Tyk versions

## Description
In order to support plugins with multiple tyk versions then was added a logic on how to load them using information such as: tyk version, operative system and architecture in which its running. The load Go Plugin logic will find first if there's an exact match with the provided path, if that file doesn't exist then it will build a path+pluginName like this:

1- Gather information like: os, tyk version and architecture
2- Obtain the clean name of the plugin
3- Obtain the directory where the plugins are stored
4-  Build a new plugin name in the format: `{plugin-dir}/{plugin-name}_{GW-version}_{OS}_{arch}.so`
5- Load the plugin using the new name

## Related Issue
https://tyktech.atlassian.net/browse/TT-4354

## Motivation and Context
Allow users to upload multiple version of a go plugin, and load it based on where its running

## How This Has Been Tested
- created go plugin following the instructions available in https://tyk.io/docs/plugins/supported-languages/golang/#plugin-development-flow
- Created api and added a plugin like:
```
 "custom_middleware": {
      "pre": [],
      "post": [
        {
          "name": "AddFooBarHeader",
          "path": "./mygoplugin/mygoplugin.so",
          "require_session": false,
          "raw_body_only": false
        }
      ],
      "post_key_auth": [],
      "auth_check": {
        "name": "",
        "path": "",
        "require_session": false,
        "raw_body_only": false
      },
      "response": [],
      "driver": "goplugin",
      "id_extractor": {
        "extract_from": "",
        "extract_with": "",
        "extractor_config": {}
      }
    },
```
- Name the plugin: `mygoplugin_v4.1.0_darwin_amd64.so`
- Curl the api: `curl -i http://tyk-gateway:8081/goplugin/get`
- In the result I can see the header Foo: Bar


**Testing the Plugin compiler**
Now the plugin compiler have the ability to generate the plugins with the name format set as explained above, in order to test it we need to create the plugin and check that the name is set as expected. The plugin compiler have the ability to receive by params the OS and Arch as 3rd and 4th param. To test, lets do:

- Create the plugin directory
- `go mod init tyk_plugin`
- `go mod edit -replace github.com/jensneuse/graphql-go-tools=github.com/TykTechnologies/graphql-go-tools@f818861b88dc
`
- `go get github.com/TykTechnologies/tyk@af8883a`
- Write the plugin following the instructions of https://tyk.io/docs/3-lts/plugins/supported-languages/golang/#write-the-plugin
- Then, to build the plugin lets use the docker image: `tykio/tyk-plugin-compiler:v4.1.0-rc3` and at this point we need to test it in 2 ways: sending the arch and os, and not sending thos params:

1- Sending OS/Arch (Skip this section as cross compilation is not working):
- Type `docker run --rm -v `pwd`:/plugin-source tykio/tyk-plugin-compiler:v4.1.0-rc5 plugin.so pluginid testOS testArch` and as output it will generate a file named: `plugin_v4.1.0_testOS_testArch.so`
The bad side of this approach is that is always required to send a pluginID.

If we attempt to send an invalid OS or arch, we will receive a message like this:
<img width="374" alt="Captura de Pantalla 2022-02-13 a la(s) 8 50 44 p  m" src="https://user-images.githubusercontent.com/4504205/153772532-008b54ab-9b42-4881-8f34-1330fb607138.png">


2- Without sending the OS/Arch:
- Type `docker run --rm -v `pwd`:/plugin-source tykio/tyk-plugin-compiler:v4.1.0-rc5 plugin.so`  as the docker container is running linux it will detect that OS and architecture and will create a file with a name like: `plugin_v4.1.0_linux_amd64.so`
- 
<img width="481" alt="Captura de Pantalla 2022-02-13 a la(s) 8 50 17 p  m" src="https://user-images.githubusercontent.com/4504205/153772282-a0389321-937f-4e37-8d80-738443ce3d47.png">


## Screenshots (if appropriate)

## Types of changes


## Checklist


      fork, don't request your `master`!
      *your branch* off *our latest `master`*.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] When updating library version must provide reason/explanation for this update.
